bgp: report reconcile errors in CiliumBGPNodeConfig conditions.

harsimran-pabla opened the pull request:
First commit will be dropped as https://github.com/cilium/cilium/pull/37420 is merged.

This change introduces error conditions in CRD CiliumBGPNodeConfig. Whenever BGP subsystem reports reconciliation error, we store some of the errors ( 5 errors per instance ) to statedb table bgp-reconcile-errors. 

Statedb table is tracked by CRD status reconcilier and it updates the error condition in the CRD. Message of the condition contains actual returned error from reconciler.

```release-note
bgp: report reconcile errors in CiliumBGPNodeConfig conditions.
```

harsimran-pabla committed:
bgp: add error condition to CiliumBGPNodeConfig CRD

Introducing reconcile error condition to CiliumBGPNodeConfig object
which can be used by users to know health of BGP control plane and if
there are any errors reported which reconciling BGP sub-system.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgp: add reconcile errors to statedb table

Adding new statedb table bgp-reconcile-errors which is populated with
first 5 reconcile errors per BGP instance.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgp: set bgpv2 mode regardless of reconcile error

BGP modev2 should be set even if reconcile error is returned. This
ensures that state is updated appropriately as reconcile error can be
transient.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgp: update crd status with error condition

Track bgp-reconcile-error statedb table and update conditions of CRD
CiliumBGPNodeConfig object.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -960,22 +977,105 @@ func (m *BGPRouterManager) reconcileBGPConfigV2(ctx context.Context,
 			//
 			// High level retry will again call the reconciliation loop as long as we return
 			// error.
+			// Error is being joined here as we may have non-abort errors previously.
```
nit: the error is now actually not joined, but appended into `reconcileErrs`, maybe we can omit this comment line now?

rastislavs commented on pkg/bgpv1/manager/tables/errors.go:
```diff
@@ -0,0 +1,92 @@
+	ReconcileErrCountPerInstance = 5
+)
+
+type ReconcileError struct {
```
Just thinking if we should put `BGP` prefix on the table type, as we do not have bgp in the package name? So only the package location makes it clear that this `ReconcileError` is specific to BGP.

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/crd_status_test.go:
```diff
@@ -5,24 +5,319 @@ package reconcilerv2
+							Type:   v2alpha1.BGPInstanceConditionReconcileError,
+							Status: metav1.ConditionTrue,
+							Reason: "BGPReconcileError",
+							Message: "ReconcileError{Instance: bgp-instance-0, ErrorID: 0, Error: error 00}\n" +
```
Wondering whether we should apply some nicer printing here - I guess the main consumers of this output are human operators, right? Maybe we could omit the ErrorID (which seems to be internal), and print something like:
```
Instance-name-0: Error1.
Instance-name-0: Error2.
Instance-name-1: Error3.
```

rastislavs reviewed (approved):
LGTM now, thanks!

youngnick commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -140,11 +155,79 @@ func NewStatusReconciler(in StatusReconcilerIn) StatusReconcilerOut {
+			observedGeneration = cond.ObservedGeneration
+			break
+		}
+	}
```
The observedGeneration is supposed to be set to the `metadata.generation` field of the record that caused the Condition to be generated, which allows the field to be a staleness check on the status for the user.

How does this `observedGeneration` get set to anything other than `0`?

Is there only one object responsible for this condition, or is it a general error across the whole set of Conditions?

Should this just be the current value of `metadata.generation` for the CiliumBGPNodeConfig object?

harsimran-pabla replied:
Thanks @youngnick, this makes sense. Please check out latest [commit](https://github.com/cilium/cilium/pull/37486/commits/c9ed92f47184e0a393eb3c0af0e7ac080b6fe731), now using metadata.generation from CiliumBGPNodeConfig object to set observed generation.

youngnick reviewed (commented):
Mostly LGTM with one question about `observedGeneration`.

harsimran-pabla committed:
bgpv2: set observed generation of CiliumBGPNodeConfig conditions.

Set observed generation of condition in the CRD based on generation of
the object.

This commit also exports mock store which helps in writing unit test for
fake object.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

youngnick reviewed (approved):
Thanks @harsimran-pabla, nice work.
